Markdown guide updates

JuanM04 opened the pull request:
- Added "Importing Markdown"
- Removed references to markdown parsers
- Updated plugins sections
- Added notes section
- Formatted

[Rendered page](https://deploy-preview-405--astro-docs-2.netlify.app/en/guides/markdown-content/)

delucis commented on src/pages/en/guides/markdown-content.md:
````diff
@@ -178,32 +177,80 @@ You can import components into your Markdown file with `setup` and use them alon
+
+```astro
+---
+// You can use `import ... from`
+import * as greatPost from '../pages/post/great-post.md';
+
+// Or even `await import()`
+const aboutUs = await import('../pages/internal/about-us.md');
````
Is there an advantage to using one of these over the other? Even if there are two ways to do this, if they’re identical, maybe it’s a good idea just to document one for simplicity? Maybe the `await import` version as it is more similar to how `Astro.glob` is used?

JuanM04 replied:
cc @FredKSchott

sarah11918 replied:
Chris' instinct is a good one. I'd also be interested to know if we have a preference to nudge people towards, and suspect that choosing the form that more closely resembles Astro.glob is a good choice for consistency.

But, I don't think that blocks this PR, and have no problems publishing this as-is and revisiting later if want to.

bholmesdev replied:
I'd advise against using await if plain ESM suits your needs. I only use await import when rendering dynamically inside the Astro templating itself.

JuanM04 replied:
I've just updated that bit, what do you think?

sarah11918 replied:
I like it!

FredKSchott replied:
personal preference -- I usually recommend this format instead:

```
---
import {Content as PromoBanner} from '../components/promoBanner.md';
--- 
{promoActive() && <PromoBanner/>}
```

As for documenting static vs. dynamic imports, we should push users towards static when we can, but it can still be nice to know that dynamic imports are supported. Maybe something more like this?

```diff
// Import some markdown. Dynamic import() is also supported!
import * as greatPost from '../pages/post/great-post.md';
- // Or even `await import()`
- const aboutUs = await import('../pages/internal/about-us.md');
```

but fwiw I like how you have it already also, so I'm +1 either way / leave the final wording up to you.

sarah11918 replied:
So... back to the original, @JuanM04 ? 😅   Either your latest version, or the original/Fred's is fine with me!

FredKSchott replied:
~~Dealer's~~ Juan's choice!

JuanM04 replied:
> Either your latest version, or the original/Fred's is fine with me!

¿Por qué no los dos?

sarah11918 replied:
muí caliente! 🔥 

(If there's any way to get a little more spacing at the bottom between the code sample and the next list item, it would look a little better.)

JuanM04 replied:
I don't know how to do that, so I moved the items around [bda5eaa](https://github.com/withastro/docs/pull/405/commits/bda5eaaa39d4c8cf9fadd846bbb241a88f72c0f9)

sarah11918 replied:
I love it even more! Winner?

delucis replied:
Looking gooooood!!!

FredKSchott replied:
LGTM!

delucis commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -178,32 +177,80 @@ You can import components into your Markdown file with `setup` and use them alon
+const aboutUs = await import('../pages/internal/about-us.md');
+
+// Lastly, you can import multiple files with Astro.glob
+const posts = await Astro.glob<Frontmatter>('../pages/post/*.md');
```
You can use `Astro.glob` without a type argument right? What do you think about presenting this first usage without it for non-Typescript users given that you cover passing a `Frontmatter` type below?

```suggestion
const posts = await Astro.glob('../pages/post/*.md');
```

delucis commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -277,40 +324,21 @@ const content = await fetch('https://raw.githubusercontent.com/withastro/docs/ma
+Astro supports third-party [remark](https://github.com/remarkjs/remark) and [rehype](https://github.com/rehypejs/rehype) plugins for Markdown. You can provide your plugins in `astro.config.mjs`.
 
-> **Note:** Enabling custom `remarkPlugins` or `rehypePlugins` removes Astro’s built-in support for the plugins previously mentioned. You must explicitly add these plugins to your `astro.config.mjs` file, if desired.
+By default, Astro comes with [GitHub-flavored Markdown](https://github.com/remarkjs/remark-gfm) and [remark-smartypants](https://github.com/silvenon/remark-smartypants) pre-enabled. Take into account that enabling custom `remarkPlugins` or `rehypePlugins` will remove these built-in plugins and you would need to explicitly add these plugins if desired.
```
The fact that this disables defaults seems important, so maybe it’s worth preserving the extra formatting to highlight it.

```suggestion
> **Note:** By default, Astro comes with [GitHub-flavored Markdown](https://github.com/remarkjs/remark-gfm) and [remark-smartypants](https://github.com/silvenon/remark-smartypants) pre-enabled. Enabling custom `remarkPlugins` or `rehypePlugins` will remove these built-in plugins and you need to explicitly add these plugins if desired.
```

delucis commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -324,7 +352,6 @@ export default {
-      // 'rehype-slug',
       // ['rehype-autolink-headings', { behavior: 'prepend'}],
     ],
   },
```
If we indent this code block (3 or 4 spaces works I think), it will nest under number 2 in the list rather than appearing as a block after the end of the `<ol>`. Same goes presumably for the text and code block that follow.

JuanM04 replied:
Done! Take a look

delucis replied:
Thanks! This is a bit of a nitpick: I would probably also indent the code block that follows, but I could understand if you don’t think that belongs under item **2.**

<img width="816" alt="image" src="https://user-images.githubusercontent.com/357379/166049286-ca93902b-fc3a-44d8-9da5-aa8c66d52fff.png">

JuanM04 replied:
Yes, it's nicer indented :smile:

delucis commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -395,9 +424,14 @@ We also suggest [diving into their theme documentation](https://github.com/shiki
 
 When using Prism, you'll need to add a stylesheet to your project for syntax highlighting. If you're just getting started and prefer to use Prism over Shiki, we suggest:
+
 1. [Setting `syntaxHighlight: 'prism'`](#choose-a-syntax-highlighter) from your `@astrojs/markdown-remark` config.
```
I’m noticing that `remark-smartypants` turns `syntaxHighlight: 'prism'` into `syntaxHighlight: ‘prism’`, which isn’t great for copy-pasting. Maybe we want to look at disabling SmartyPants for the docs? I don’t think it adds anything essential or can it be configured to ignore code snippets?

JuanM04 replied:
I think this will be fixed by #404

delucis commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -395,9 +424,14 @@ We also suggest [diving into their theme documentation](https://github.com/shiki
+
+## Notes
+
+- Astro will add autogenerated ids to all the headings automatically using [github-slugger](https://github.com/Flet/github-slugger). If a custom id is specified, it won't be overriden.
```
Not sure if this clarification is necessary given this is the Markdown page, but probably wouldn’t hurt.

```suggestion
- Astro will add autogenerated ids to all headings in Markdown files automatically using [github-slugger](https://github.com/Flet/github-slugger). If a custom id is specified, it won't be overriden.
```

delucis reviewed (commented):
Looks great @JuanM04! Very thorough. I’ve made a few suggestions/left a few questions.

sarah11918 commented:
@JuanM04 This is amazingly helpful! Thank you for this!  I'm going to read through it a bit later, but I just wanted to add here that we also happen to need to document the Markdown Content component as per issue #358 

If you, or anyone working on this wants to add some "content" about that in here, or suggest where/what in comments, that would be really helpful, too!

JuanM04 commented:
@sarah11918 I've added a bit more of documentation about `Content`. I don't know exactly how Content works, so I don't know what else to write.

Also, I've modified the Notes at the end

sarah11918 commented on src/pages/en/guides/markdown-content.md:
````diff
@@ -178,32 +177,72 @@ You can import components into your Markdown file with `setup` and use them alon
+```
 
--->
+Each one of the examples above return an (or multiple) `MarkdownInstance`, an object with the following properties:
````
```suggestion
Each one of the examples above return a (or multiple) `MarkdownInstance`, an object with the following properties:
```

sarah11918 replied:
I *think* this would be correct, even though it might seem to go against the prononciation rule.

Another way around this could be something ilke, "Each Markdown file imported returns a `MarkdownInstance`..."

sarah11918 commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -178,32 +177,72 @@ You can import components into your Markdown file with `setup` and use them alon
+Each one of the examples above return an (or multiple) `MarkdownInstance`, an object with the following properties:
+
+- `frontmatter`: Any data specified in this file's YAML frontmatter.
+- `file`: The absolute path of this file (like `/home/user/projects/.../file.md`).
```
```suggestion
- `file`: The absolute path of this file (e.g. `/home/user/projects/.../file.md`).
```

I think we more often tend to use "e.g." than "like" in general.

sarah11918 commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -178,32 +177,72 @@ You can import components into your Markdown file with `setup` and use them alon
+
+- `frontmatter`: Any data specified in this file's YAML frontmatter.
+- `file`: The absolute path of this file (like `/home/user/projects/.../file.md`).
+- `url`: If it's a page, URL of the page (like `/en/guides/markdown-content`).
```
```suggestion
- `url`: If it's a page, URL of the page (e.g. `/en/guides/markdown-content`).
```

sarah11918 commented on src/pages/en/guides/markdown-content.md:
````diff
@@ -277,59 +316,40 @@ const content = await fetch('https://raw.githubusercontent.com/withastro/docs/ma
-  },
-};
-```
+You can customize the Markdown parsing by modifing your `astro.config.mjs`. [Here you can read the full reference](/en/reference/configuration-reference/#markdown-options).
````
```suggestion
You can customize your Markdown parsing by modifing your `astro.config.mjs`. [Here you can read the full reference](/en/reference/configuration-reference/#markdown-options).
```

sarah11918 commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -395,9 +417,14 @@ We also suggest [diving into their theme documentation](https://github.com/shiki
 
 You can also visit the [list of languages supported by Prism](https://prismjs.com/#supported-languages) for options and usage.
+
+## Notes
```
Just checking: a header at this level is showing up in the right hand sidebar, so is this the best description to use? Given that this seems to be just one note, can the header be more descriptive? e.g "Heading IDs"?

And then, I don't think it's necessary to use a bullet point for this. I don't think it benefits from the point and the indentation.

JuanM04 replied:
Since I didn't know where to put this information, I thought of a misc section. Maybe "Technical Details" would be a better name

JuanM04 replied:
I've added a note about `<Markdown />` and SSR. What do you think? Do you like this approach or should I search another place to put them?

sarah11918 replied:
Fred has an entire new PR #411 with warnings splashed all over the page about the Markdown component! So, I don't think we need it here.  😄  Take a look at that one, then probably I think just remove your SSR bit from here. Once THIS pr is merged, we can merge Fred's.

sarah11918 commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -395,9 +417,14 @@ We also suggest [diving into their theme documentation](https://github.com/shiki
+
+## Notes
+
+- Astro will add autogenerated ids to all headings in Markdown files automatically using [github-slugger](https://github.com/Flet/github-slugger) — if a custom id is specified, it won't be overriden. These ids will be added _after_ all the other plugins are executed, so if you have a plugin like `rehype-toc` that needs ids, you should add your own slugging plugin (like `rehype-slug`).
```
```suggestion
Astro will add autogenerated ids to all headings in Markdown files automatically using [github-slugger](https://github.com/Flet/github-slugger). But, if a custom id is specified, it won't be overriden.

  These ids will be added _after_ all the other plugins are executed, so if you have a plugin like `rehype-toc` that needs ids, you should add your own slugging plugin (like `rehype-slug`).
```

I would just build this out like another normal section.

sarah11918 replied:
(Pssst: I still like this change with a more descriptive title IF there's only stuff about Markdown ids for now and you don't have more stuff you think you'll want to add here soon. But I'm happy to let you make the final decision!)

JuanM04 replied:
I've update it :smile:

sarah11918 commented:
@JuanM04 after you take a look at PR #411 so you know what you do NOT need to mention at the end re: SSR/Markdown component, I'm happy for you to merge any time you're ready!  Thank you again for this amazing work! 🥳

FredKSchott commented on src/pages/en/guides/markdown-content.md:
```diff
@@ -395,9 +417,16 @@ We also suggest [diving into their theme documentation](https://github.com/shiki
 
 You can also visit the [list of languages supported by Prism](https://prismjs.com/#supported-languages) for options and usage.
+
+## Heading IDs
```
nit: I'd move this up the page, closer to where we describe other features like "Drafts"

JuanM04 replied:
Moved it, is that a good spot?

FredKSchott reviewed (approved):
LGTM!

JuanM04 commented:
I think there are no more corrections. Speak now or forever hold your peace

sarah11918 commented:
Go, go, @JuanM04 ! We need to merge this so we can move on to #411 and get this page crossed off my list! ;)

Great work, and thanks for such dedication. 🎉
